Replace parse_header with parse_header_parameters

jaap3 opened the pull request:
Add a backwards compatibility shim for Django versions that have no (or an incompatible)
django.utils.http.parse_header_parameters implementation.

This is an alternative fix to the one proposed in #8543, which also attempts to fix #8540.

While making this PR I noticed that get_encoded_filename is not necesarry anymore as this is already handled correctly in Django ~1.8 (https://github.com/django/django/commit/b42e5ca058178d67027bf66d37d00ade635b4c26)

carltongibson commented on rest_framework/compat.py:
```diff
@@ -152,6 +152,25 @@ def md_filter_add_syntax_highlight(md):
         return False
 
 
+try:
+    # Django <= 4.1, create a compatibility wrapper for parse_header_parameters
+    from django.http.multipartparser import parse_header
+
+    def parse_header_parameters(line):
+        main_value_pair, params = parse_header(line.encode())
+        return main_value_pair, {
+            k: v if isinstance(v, str) else v.decode()
+            for k, v in params.items()
+        }
+
+except ImportError:
+    # Django 4.2+, use the new built-in function
+    # Note: Django 4.1 also has an implementation of parse_header_parameters
+    #       which is slightly different from the one in 4.2
+    from django.utils.http import parse_header_parameters
+    parse_header_parameters = parse_header_parameters
+
+
```
Normally we'd import the future version, falling back to the shim, so... 

```python
try: 
    from django.utils.http import parse_header_parameters
except ImportError:
   from django.http.multipartparser import parse_header
 
   def parse_header_parameters(...):
       ...

jaap3 replied:
Yes, but as I noted in the comment:

# Note: Django 4.1 also has an implementation of parse_header_parameters
#       which is slightly different from the one in 4.2

jaap3 replied:
I could do an explicit Django version check here, (`if django.VERSION > (4, 1):`) if you'd prefer that

carltongibson replied:
I'm used to reading the new version first, so yes I think I do prefer it that way round. Thanks!

carltongibson reviewed (commented):
Hey @jaap3 — thanks for this, looks good. 

Just one initial comment on the compat change... 

@shaib's comment from https://github.com/encode/django-rest-framework/pull/8543#discussion_r919068186 about **WHY** we `decode()` in the shim may be worth adding, for later comprehension too.

jaap3 commented:
I've addressed the review comments.

BTW, as this issue is only really present in django's main branch it's hard to spot the breakage in the CI, as tox is setup to allow failure for that particular "version". Looking at the results here https://github.com/encode/django-rest-framework/runs/7320233994?check_suite_focus=true#step:6:363 you'll see that djangomain now passes, when it didn't before (i.e. the most recent run for master: https://github.com/encode/django-rest-framework/runs/7094624536?check_suite_focus=true#step:6:363)

carltongibson commented on rest_framework/compat.py:
```diff
@@ -152,6 +153,32 @@ def md_filter_add_syntax_highlight(md):
+    #       which is slightly different from the one in 4.2, it needs
+    #       the compatibility shim as well.
+    from django.utils.http import parse_header_parameters
+    parse_header_parameters = parse_header_parameters
```
This line isn't needed I think.

carltongibson replied:
Hmm. OK. Let me have a look... 🤔

Thanks!

carltongibson reviewed (commented):
OK, looks great. 

Just the one query...

carltongibson commented:
> ...as tox is setup to allow failure for that particular "version". 

Yes... We didn't yet get to failing on breaks again django/django main, but probably should at some point. //cc @tomchristie

carltongibson commented:
Hi @jaap3... 

> flake8 yells at me if I don't do this

Hmmm 🤔 Not sure what's going on there. Passes locally for me. 

Also on CI https://github.com/encode/django-rest-framework/runs/7339596012?check_suite_focus=true

I'll finish this in #8556, since the _Allow maintainers to make edits_ box seems to be unchecked. 

Thanks for the input. Great stuff! 🏅
